### PR TITLE
Low: libcrmcommon: Fix handling node=NULL in pcmk__attrd_api_query.

### PR DESCRIPTION
**Pull request by clumens**
According to the header file, if node is NULL, pcmk__attrd_api_query should query the value of the given attribute on all cluster nodes. This is also what the server expects and how attrd_updater is supposed to work.

However, pcmk__attrd_api_query does not correctly handle a NULL here. Instead, it will call pcmk__node_attr_target.  That function takes NULL to mean that it should probe the local cluster node for its name, returning that to pcmk__attrd_api_query.  If it returns non-NULL, that value will then be put into the XML IPC call which means the server will only return the value for that node.

In testing this was usually fine.  However, in pratice, the methods pcmk__node_attr_target uses to figure out the local cluster node name involves checking the OCF_RESKEY_CRM_meta_on_node environment variable among others.

This variable was never set in testing, but can be set in the real world.  This leads to circumstances where the user did "attrd_updater -QA" expecting to get the values on all nodes, but instead only got the value on the local cluster node.

In pacemaker-2.1.4 and prior, pcmk__node_attr_target was simply never called if the node was NULL but was called otherwise.  It's possible that we should never be calling this function for queries, but in the interests of restoring previous behavior, we're just going to skip it in the NULL case here.

Regression in 2.1.5 introduced by eb20a65577

### CODE REVIEW

**Comment by clumens** [comment]
@nrwahl2 Can you give this a test?

**Comment by nrwahl2** [comment]
I'm also of the opinion that `pcmk__node_attr_target()` probably shouldn't be called at all for queries, but I agree that we should preserve that behavior in case anyone depends on it.

I'd like to add a `@COMPAT` comment to revisit that later. Otherwise it looks good. I'm building now to test. I don't expect any issues (you've tested, regression passes, and it's fine by inspection), but I'll comment if anything pops up.

**Review by nrwahl2 — APPROVED** [review]

**Comment by nrwahl2** [comment]
> I'm also of the opinion that `pcmk__node_attr_target()` probably shouldn't be called at all for queries

I'm not confident in that though. If we did specify a node (not `NULL`), it might be reasonable to query the same node that we would use to set the attribute when the same node name is given. Thinking especially of things like containers, which `pcmk__node_attr_target()` handles. So maybe it would be better to just update the doxygen for `pcmk__node_attr_target()`. The merits of using of `on_node` are somewhat less clear. It's worth giving some additional thought.

None of that affects the fix here, which I still think is correct. I just tested it successfully with `OCF_RESKEY_CRM_meta_on_node`.

**Review by kgaillot — COMMENTED** [review]
The existence of the --all option suggests that if it is *not* specified, then an unspecified node *should* be mapped to the local node. It looks like that was the intent when it was added in 3c998292b. I'm thinking we need another `pcmk__node_attr_` option to pass along query all to api_query().

**Comment by nrwahl2** [comment]
> The existence of the --all option suggests that if it is not specified, then an unspecified node should be mapped to the local node.

That sounds reasonable, if that preserves previous behavior. I mentioned to Chris in chat that the `query_all` variable is redundant right now.

> to the local node

A lot of times it *isn't*. What made this difficult to diagnose/reproduce is that running `attrd_updater -n attr_name -Q` from the command line doesn't map to the local node, because `OCF_RESKEY_CRM_meta_on_node` is not set.

It maps to the local node from within a resource agent. From other scripts or from the bare command line, if `-A` is not specified, we get the attribute from all nodes (unless `container-attribute-target` is set). So even the question of preserving existing behavior gets sticky. Users might be relying on the fact that "not specifying `-A`" can get all nodes (if they're running `attrd_updater` in a custom script) or map to the local node (if they're running `attrd_updater` in a resource agent).

**Comment by clumens** [comment]
I think this latest patch handles all cases - pcmk__attrd_api_query can query the local node if NULL is given, the given node, or all nodes if the flag is given.  If you give attrd_updater -QA it queries all nodes, if you give --node= it queries that node, and if you give neither but OCF_RESKEY_CRM_meta_on_node is set in the environment it queries that node.

File: include/crm/common/ipc_attrd_internal.h [hunk 1]
```diff
@@ -110,10 +110,13 @@ int pcmk__attrd_api_purge(pcmk_ipc_api_t *api, const char *node);
...
  * \param[in,out] api           Connection to pacemaker-attrd
  * \param[in]     node          Look up the attribute for this node
- *                              (or NULL for all nodes)
+ *                              (or NULL for the local node)
  * \param[in]     name          Attribute name
  * \param[in]     options       Bitmask of pcmk__node_attr_opts
  *
+ * \note Passing pcmk__node_attr_query_all will cause the function to query
+ *       the value of \p name on all nodes, regardless of the value of \p node.
+ *
```
Comment (nrwahl2) [inline include/crm/common/ipc_attrd_internal.h:119]: copyright in this file

File: tools/attrd_updater.c [hunk 2]
```diff
@@ -431,6 +431,7 @@ static int
 send_attrd_query(pcmk__output_t *out, const char *attr_name,
                  const char *attr_node, gboolean query_all)
 {
+    uint32_t options = 0;
```
Comment (nrwahl2) [inline tools/attrd_updater.c:434]: copyright in this file

Comment (clumens) [reply tools/attrd_updater.c:434]: I think you meant in lib/common/ipc_attrd.c.  Fixed.

File: include/crm/common/ipc_attrd_internal.h (hunk 1 above)
Comment (clumens) [reply include/crm/common/ipc_attrd_internal.h:119]: Fixed.

File: tools/attrd_updater.c (hunk 2 above)
Comment (nrwahl2) [reply tools/attrd_updater.c:434]: Whoops, I scrolled too far down.

**Comment by clumens** [comment]
> Aside - I was hoping we'd be able to use the new cts-attrd to test this, but it kind of requires multiple nodes to really test, and there's no way to set that up with cts-attrd.

Actually, weirdly, attrd_updater will let you set attributes on nodes that do not exist (I suppose in case that node joins in the future?):

```
[root@rhel8-cluster-1 ~]# attrd_updater --name AAA -U 111 --node rhel8-cluster-4
[root@rhel8-cluster-1 ~]# echo $?
0
[root@rhel8-cluster-1 ~]# attrd_updater --name AAA -QA
name="AAA" host="rhel8-cluster-2" value="111"
name="AAA" host="rhel8-cluster-1" value="111"
name="AAA" host="rhel8-cluster-4" value="111"
```

I assure you there is no rhel8-cluster-4.  So I could use this quirk to write a test for this.

**Comment by kgaillot** [comment]
> Actually, weirdly, attrd_updater will let you set attributes on nodes that do not exist (I suppose in case that node joins in the future?):

Yes, that's intentional. For example, you can set attributes for a new node before adding it, so that when it joins, any attribute-based rules immediately apply correctly. In Pacemaker 1, nodes could also join via a broadcast mechanism, which meant that the entire set of nodes was never known reliably.

**Comment by kgaillot** [comment]
> I think this latest patch handles all cases - pcmk__attrd_api_query can query the local node if NULL is given, the given node, or all nodes if the flag is given. If you give attrd_updater -QA it queries all nodes, if you give --node= it queries that node, and if you give neither but OCF_RESKEY_CRM_meta_on_node is set in the environment it queries that node.

It looks like we never really paid sufficient attention to how query worked without a node :-/

Technically the help doesn't specify what happens when --query is given without --node. The --node help only mentions setting attributes, not querying. The ideal would be for query to pass the string "localhost" when node is not given (which tells pacemaker-attrd to substitute the local node name), but given how long the behavior's been how it is, we should wait until the next minor release to do that. For now we can keep the current behavior.

File: tools/attrd_updater.c (hunk 2 above)
Comment (kgaillot) [inline tools/attrd_updater.c:434]: pcmk__node_attr_none

Also, regressions should use High in the commit message

File: python/pacemaker/_cts/corosync.py [hunk 3]
```diff
@@ -11,6 +11,8 @@
 from pacemaker.buildoptions import BuildOptions
 from pacemaker._cts.process import killall, stdout_from_command
 
+corosync_cfg_file = BuildOptions.COROSYNC_CONFIG_DIR + "/corosync.conf"
```
Comment (kgaillot) [inline python/pacemaker/_cts/corosync.py:14]: Bah, this is actually a bug in configure.ac :-/ It is supposed to be the file path. It looks like the only use of the variable is in cts/lab/CIB.py, and it's probably broken when corosync provides pkg-config variables. configure.ac correctly uses the file path in all other cases.

File: cts/cts-attrd.in [hunk 4]
```diff
@@ -183,6 +183,33 @@ class AttributeTests(Tests):
...
+        """ Add tests that set and query an attribute across multiple nodes """
+
+        # NOTE:  These tests make use of the fact that nothing in attrd actually
+        # cares about whether a node exists when you set or query an attribute.
+        # It just keeps creating new hash tables for each node you ask it about.
+
+        test = self.new_test("multi_query_1",
+                             "Query an attribute set across multiple nodes")
+        test.add_cmd("attrd_updater", "--name AAA -U 111 --node cluster1 --output-as=xml")
+        test.add_cmd("attrd_updater", "--name AAA -U 111 --node cluster2 --output-as=xml")
```
Comment (kgaillot) [inline cts/cts-attrd.in:196]: use different values to verify that it's only setting the attribute for the one node

File: tools/attrd_updater.c (hunk 2 above)
Comment (clumens) [reply tools/attrd_updater.c:434]: Fixed.

File: cts/cts-attrd.in (hunk 4 above)
Comment (clumens) [reply cts/cts-attrd.in:196]: Good idea, fixed.

File: python/pacemaker/_cts/corosync.py (hunk 3 above)
Comment (clumens) [reply python/pacemaker/_cts/corosync.py]: Fixed.

**Pull request merged**